sled-agent/src/rack_setup/service.rs

```diff
@@ -867,6 +859,33 @@ impl ServiceInner {
                 .await?;
         }
 
+        // Save the relevant network config in the bootstore. We want this to
+        // happen before we `initialize_sleds` so each scrimlet (including us)
+        // can use its normal boot path of "read network config for our switch
+        // from the bootstore".
+        if let Some(rack_network_config) = &config.rack_network_config {
+            let early_network_config = EarlyNetworkConfig {
+                generation: 1,
+                rack_subnet: config.rack_subnet,
+                ntp_servers: config.ntp_servers.clone(),
+                rack_network_config: rack_network_config.clone(),
+            };
+            info!(self.log, "Writing Rack Network Configuration to bootstore");
+            bootstore
+                .update_network_config(early_network_config.into())
+                .await?;
+            // TODO: Wait for bootstore to sync to all `plan.sleds`? Every
```

Comment:
One thing I just realized here is that without `config.rack_network_config` the sled-agent loop that waits for the bootstore network config will never complete. To satisfy this condition, we'll need to either wait for bootstore to sync to all plan.sleds or wrap the `EarlyNetworkConfig` in an `Option` so we don't use it.

I'm not really sure why we do this check here either? Do we always expect to have a `rack_network_config` form this point on? Can we just go ahead and remove the conditional? Does this work with single gimlets?

If we remove it, we definitely will not need to do the wait for sync here, and I'd prefer not to add that code to the bootstore right now.

CC @internet-diglett @smklein @rcgoodfellow

Reply:
Hm, right, if non-racked gimlet environments run without a `rack_network_config` we should change sled-agent's check for sure. Should we store a `NetworkConfig` with an `Option<RackNetworkConfig>` inside bootstore to tell the difference between "no network config sync'd yet" and "we intentionally sync'd no network config" (e.g., for bench gimlets)?

Reply:
Yeah, I think storing an `Option<RackNetworkConfig>` is the right thing to do in this case.

Reply:
IIUC this code is mostly to support development environments where a configuration file is provided directly to the sled agent.

Reply:
Switched to storing `Option<RackNetworkConfig>` to continue to support dev envs with no uplinks in [298acb4](https://github.com/oxidecomputer/omicron/pull/3765/commits/298acb4b495540056415a68bea340c223192b10a)

Reply:
> IIUC this code is mostly to support development environments where a configuration file is provided directly to the sled agent.

Yep, that and some of the tests IIRC.